### DISCUSSION TITLE
트랜잭션에 대한 설명과 특성(ACID)

### DISCUSSION THREAD

**Discussion by J-I-H-O** [2024-woowacourse-study/level-interview#190] [Database]
.

**Comment by jimi567** [comment]
트랜잭션 : 데이터베이스의 상태를 변화시키기 해서 수행하는 작업의 단위

* Commit : 하나의 트랜잭션이 성공적으로 끝났고, 데이터베이스가 일관성있는 상태에 있을 때, 
* Rollback : 하나의 트랜잭션에서 트랜잭션 처리가 비정상적으로 종료되어 트랜잭션의 원자성이 깨진경우, 트랜잭션을 처음부터 다시 시작하거나, 트랜잭션의 부분적으로만 연산된 결과를 취소시킴

트랜잭션의 특성 ACID

Atomicity(원자성) :   트랜잭션이 데이터베이스에 모두 반영되던가, 아니면 전혀 반영되지 않아야 한다
Consistency(일관성) : 트랜잭션의 작업 처리 결과가 항상 일관성이 있어야 한다
Isolation (독립성) :  어떤 하나의 트랜잭션이라도, 다른 트랜잭션의 연산에 끼어들 수 없다
Durability(지속성) : 트랜잭션이 성공적으로 완료됐을 경우, 결과는 영구적으로 반영되어야 한다

**Comment by GIVEN53** [comment]
트랜잭션은 데이터베이스 관리 시스템(DBMS)에서 하나의 논리적 작업 단위를 형성하는 일련의 데이터 조작 작업입니다. 트랜잭션은 일관된 데이터베이스 상태를 유지하고, 여러 작업을 하나의 작업처럼 처리하여 시스템의 신뢰성을 높입니다.

### Atomicity (원자성)
모든 작업이 완벽하게 수행되거나 전혀 수행되지 않음을 보장합니다. 트랜잭션 내의 모든 작업은 하나의 단위로 실행되며, 어느 하나라도 실패하면 전체 트랜잭션이 실패하고 데이터베이스 상태는 트랜잭션 시작 전으로 되돌아갑니다.

### Consistency (일관성)
트랜잭션이 성공적으로 완료되면 데이터베이스는 일관된 상태로 유지됩니다. 이는 트랜잭션이 데이터베이스의 무결성을 해치지 않음을 의미하며, 데이터베이스의 제약 조건, 규칙 및 데이터 무결성 조건을 모두 만족시킵니다.

### Isolation (격리성)
트랜잭션이 독립적으로 수행되어 다른 트랜잭션의 중간 상태를 볼 수 없음을 보장합니다. 즉, 각 트랜잭션은 다른 트랜잭션의 영향을 받지 않고 독립적으로 실행됩니다. 이는 트랜잭션 간의 간섭을 방지하고 데이터 일관성을 유지합니다.

### Durability (지속성)
트랜잭션이 성공적으로 완료된 후, 그 결과는 영구적으로 저장되며 시스템 오류가 발생하더라도 손실되지 않음을 보장합니다. 이는 트랜잭션이 완료된 후의 변경 사항이 안정적인 저장 매체에 기록됨을 의미합니다.

### 트랜잭션의 동작 과정
시작: 트랜잭션이 시작됩니다. 이 시점부터 데이터베이스 작업이 트랜잭션 단위로 수행됩니다.
작업 수행: 데이터베이스에 여러 작업(삽입, 갱신, 삭제 등)을 수행합니다.
커밋 (Commit): 트랜잭션이 성공적으로 완료되어 모든 변경 사항이 데이터베이스에 영구적으로 적용됩니다.
롤백 (Rollback): 트랜잭션이 실패하면, 트랜잭션이 시작된 시점으로 되돌아가 모든 변경 사항이 취소됩니다.

**Comment by le2sky** [comment]
DBMS 상호작용 단위
논리적인 작업 단위
데이터 무결성
커밋과 롤백 연산이 존재
원자성, 일관성, 격리성, 지속성
